### DISCUSSION TITLE
Reopen Firefox windows on the right i3 workspaces

### DISCUSSION THREAD

**Discussion by yurikhan** [i3/i3#5964] [Show and tell]
TL;DR: I made a Firefox [extension](https://github.com/yurikhan/firefox-i3-workspaces) that keeps track which window belongs to which workspace.

# Problem statement

I have lots of tabs in many windows over several workspaces across a few monitors, and I rely on windows staying where I put them. For example, the window dedicated to workplace chat server belongs on the workspace fixed to my side screen, and windows specific to each task I perform move through workspaces `3 wip` → `4 review` → `5 done` as their status changes.

When I need to restart the browser or reboot the system, Firefox is set up to restore tabs and windows from the last session. However, it dumps all windows on the current workspace. So I used to have to spend the first five minutes putting windows where they were supposed to be.

# Non-solutions

Firefox has code that is supposed to put windows on the workspaces they were last on, but it is specifically disabled if running on i3, for some technical reason that I’m not ready to dive into.

i3 has layout saving, but I found it impossible to craft reliable swallow criteria. All Firefox windows have the same `class`, `instance`, and `window_role` properties, so not unique. And the title reflects the title of the page in the currently active tab, so not stable. I would have to invoke the `i3-save-layout` script each time before quitting Firefox and massage it into something usable with `append_layout`.

# My approach

A Firefox extension tacks a UUID on every window and keeps track of the i3 workspace it was last seen on, in a way that survives restarts of Firefox, i3, and the whole system. The extension collaborates with a Python script that uses the i3 IPC API to supply it with necessary information, advise it of any relevant changes, and move windows on its behalf.

I have been running this for a few weeks and find it stable enough for me. I’m sure someone devoted would find ways to break it, but probably not in a critical way.

# Conditions

The extension and host script are licensed under GPL 3.0.

Bug reports will be considered, especially if accompanied by a repro. For feature requests, be prepared to bring patches ;)
